src/mirth/elab.mth

```diff
@@ -293,45 +293,56 @@ def(elab-arrow-step-buffer!, Arrow Buffer -- Arrow +Elab,
 def(elab-arrow-step-external!, Arrow External -- Arrow +Elab,
     dup2 swap arrow-op-external!
     elab-token@ token-args-0 # high order not yet supported
-    elab-external-sig! elab-arrow-apply-sig-type!)
+    elab-external-sig! elab-arrow-freshen-apply-sig-type!)
 
 def(elab-arrow-step-word!, Arrow Word -- Arrow +Elab,
     dup2 swap arrow-op-word!
     elab-token@ token-args-0 # high order not yet supported
-    elab-word-sig! elab-arrow-apply-sig-type!)
+    elab-word-sig! elab-arrow-freshen-apply-sig-type!)
 
 def(elab-arrow-step-stack-assertion!, Arrow Token -- Arrow +Elab,
+    tuck
     token-brace-arg
-    dip(elab-ctx@)
-    elab-type-stack! drop nip
+    dip(elab-ctx@ dup)
+    elab-type-stack! drop
+    rotr swap dup2 = if(drop2,
+        "stack assertion introduced new type variables:" elab-emit-warning!
+        "\tbefore: " str-trace! ctx-trace! trace-ln!
+        "\tafter:  " str-trace! ctx-trace! trace-ln!
+        drop
+        TYPE_ERROR
+    )
     dup morphism-type-new! TMorphism
     elab-arrow-apply-sig-type!
+    dip(elab-token!)
```

Comment:
oh right. This is probably because elab-type-stack! can currently change the token, but it shouldn't. Oops. I think `elab-type-unify!` is at fault (it's a terrible hack).